File: person_task_manager/server/src/ui/controllers/project.controller.ts
import { NextFunction, Request } from "express";
import { projectService } from "../../core/services/project.service";
import { IResponse } from "../../core/common/response";
import { plainToInstance } from "class-transformer";
import { ProjectRequestDto } from "../../core/domain/dtos/project.dto";

class ProjectController {

    constructor() { }
    
    async listAllProjects(req: Request, next: NextFunction): Promise<IResponse | undefined> {
        try {
            const userId = Number(req.params.userId);
            return await projectService.getAllProjects(userId);
        } catch (err) {
            next(err);
        }
    }

    async getProjectById(req: Request, next: NextFunction): Promise<IResponse | undefined> {
        try {
            const id = req.params.id;
            return await projectService.getProject(id);
        } catch (err) {
            next(err);
        }
    }

    async createProject(req: Request, next: NextFunction): Promise<IResponse | undefined> {
        try {
            const createProjectObjectDto = plainToInstance(ProjectRequestDto, req.body);
            return await projectService.createProject(createProjectObjectDto);
        } catch (err) {
            next(err);
        }
    }

    async updateProject(req: Request, next: NextFunction): Promise<IResponse | undefined> {
        try {
            const id = req.params.id;
            const updateProjectObjectDto = plainToInstance(ProjectRequestDto, req.body);
            return await projectService.updateProject(id, updateProjectObjectDto);
        } catch (err) {
            next(err);
        }
    }

    async deleteProject(req: Request, next: NextFunction): Promise<IResponse | undefined> {
        try {
            const id = req.params.id;
            return await projectService.deleteProject(id);
        } catch (err) {
            next(err);
        }
    }

    async getGroupTasksInProject(req: Request, next: NextFunction): Promise<IResponse | undefined> {
        try {
            const id = req.params.id;
            return await projectService.getGroupTasksInProject(id);
        } catch (err) {
            next(err);
        }
    }

    async getProjectsWithGroupTasks(req: Request, next: NextFunction): Promise<IResponse | undefined> {
        try {
            const userId = Number(req.params.userId);
            return await projectService.getProjectsWithGroupTasks(userId);
        } catch (err) {
            next(err);
        }
    }

    async updateProjectName(req: Request, next: NextFunction): Promise<IResponse | undefined> {
        try {
            const id = req.params.id;
            return await projectService.updateProjectName(id, req.body.name);
        } catch (err) {
            next(err);
        }
    }

    async updateProjectColor(req: Request, next: NextFunction): Promise<IResponse | undefined> {
        try {
            const id = req.params.id;
            return await projectService.updateProjectColor(id, req.body.color);
        } catch (err) {
            next(err);
        }
    }

    async archiveProject(req: Request, next: NextFunction): Promise<IResponse | undefined> {
        try {
            const id = req.params.id;
            return await projectService.archiveProject(id);
        } catch (err) {
            next(err);
        }
    }

    async enableProject(req: Request, next: NextFunction): Promise<IResponse | undefined> {
        try {
            const id = req.params.id;
            return await projectService.enableProject(id);
        } catch (err) {
            next(err);
        }
    }
}

export const projectController = new ProjectController();

File: middleware_loader/ui/controller_services/task_manager/project_controller.go
package controller_services

import (
	"fmt"
	base_dtos "middleware_loader/core/domain/dtos/base"
	"middleware_loader/core/middleware"
	mapper "middleware_loader/core/port/mapper/request"
	services "middleware_loader/core/services/task_manager"
	"middleware_loader/infrastructure/graph/model"
	"middleware_loader/kernel/utils"
	"middleware_loader/ui/controller_services/controller_utils"
	"net/http"

	"github.com/go-chi/chi"
)

func ListAll(w http.ResponseWriter, r *http.Request, projectService *services.ProjectService) {
	userId := fmt.Sprintf("%.0f", r.Context().Value(middleware.ContextKeyUserId))
	input := mapper.GetId(userId)

	graphqlQueryModel := []base_dtos.GraphQLQuery{}
	graphqlQueryModel = append(graphqlQueryModel, base_dtos.GraphQLQuery{FunctionName: "listAllProjectsByUserId", QueryInput: input, QueryOutput: model.Project{}})
	graphqlQuery := utils.GenerateGraphQLQueryWithMultipleFunction("query", graphqlQueryModel)

	utils.ConnectToGraphQLServer(w, graphqlQuery)

}

func GetById(w http.ResponseWriter, r *http.Request, projectService *services.ProjectService) {
	projectId := chi.URLParam(r, "id")
	input := mapper.GetId(projectId)

	graphqlQueryModel := []base_dtos.GraphQLQuery{}
	graphqlQueryModel = append(graphqlQueryModel, base_dtos.GraphQLQuery{FunctionName: "getProjectById", QueryInput: input, QueryOutput: model.Project{}})
	graphqlQuery := utils.GenerateGraphQLQueryWithMultipleFunction("query", graphqlQueryModel)

	utils.ConnectToGraphQLServer(w, graphqlQuery)
}

func CreateProject(w http.ResponseWriter, r *http.Request, projectService *services.ProjectService) {
	var body map[string]interface{}
	body, err := controller_utils.MappingBody(w, r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	userId := fmt.Sprintf("%.0f", r.Context().Value(middleware.ContextKeyUserId))
	input := mapper.CreateProjectRequestDTOMapper(body, userId)

	graphqlQueryModel := []base_dtos.GraphQLQuery{}
	graphqlQueryModel = append(graphqlQueryModel, base_dtos.GraphQLQuery{FunctionName: "createProject", QueryInput: input, QueryOutput: model.Project{}})
	graphqlQuery := utils.GenerateGraphQLQueryWithMultipleFunction("mutation", graphqlQueryModel)

	utils.ConnectToGraphQLServer(w, graphqlQuery)

}

func UpdateProject(w http.ResponseWriter, r *http.Request, projectService *services.ProjectService) {
	var body map[string]interface{}
	body, err := controller_utils.MappingBody(w, r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	projectId := chi.URLParam(r, "id")
	userId := fmt.Sprintf("%.0f", r.Context().Value(middleware.ContextKeyUserId))
	input := mapper.UpdateProjectRequestDTOMapper(body, projectId, userId)

	graphqlQueryModel := []base_dtos.GraphQLQuery{}
	graphqlQueryModel = append(graphqlQueryModel, base_dtos.GraphQLQuery{FunctionName: "updateProject", QueryInput: input, QueryOutput: model.Project{}})
	graphqlQuery := utils.GenerateGraphQLQueryWithMultipleFunction("mutation", graphqlQueryModel)

	utils.ConnectToGraphQLServer(w, graphqlQuery)
}

func DeleteProject(w http.ResponseWriter, r *http.Request, projectService *services.ProjectService) {
	projectId := chi.URLParam(r, "id")
	input := mapper.GetId(projectId)

	graphqlQueryModel := []base_dtos.GraphQLQuery{}
	graphqlQueryModel = append(graphqlQueryModel, base_dtos.GraphQLQuery{FunctionName: "deleteProject", QueryInput: input, QueryOutput: model.Project{}})
	graphqlQuery := utils.GenerateGraphQLQueryWithMultipleFunction("mutation", graphqlQueryModel)

	utils.ConnectToGraphQLServer(w, graphqlQuery)
}

func GetGroupTasksInProject(w http.ResponseWriter, r *http.Request, projectService *services.ProjectService) {
	projectId := chi.URLParam(r, "id")

	input := mapper.GetId(projectId)

	graphqlQueryModel := []base_dtos.GraphQLQuery{}
	graphqlQueryModel = append(graphqlQueryModel, base_dtos.GraphQLQuery{FunctionName: "getGroupTasksInProject", QueryInput: input, QueryOutput: model.GroupTask{}})
	graphqlQuery := utils.GenerateGraphQLQueryWithMultipleFunction("query", graphqlQueryModel)

	utils.ConnectToGraphQLServer(w, graphqlQuery)
}

func UpdateProjectName(w http.ResponseWriter, r *http.Request, projectService *services.ProjectService) {
	var body map[string]interface{}
	body, err := controller_utils.MappingBody(w, r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	projectId := chi.URLParam(r, "id")
	input := mapper.UpdateProjectNameRequestDTOMapper(body, projectId)

	graphqlQueryModel := []base_dtos.GraphQLQuery{}
	graphqlQueryModel = append(graphqlQueryModel, base_dtos.GraphQLQuery{FunctionName: "updateProjectName", QueryInput: input, QueryOutput: model.Project{}})
	graphqlQuery := utils.GenerateGraphQLQueryWithMultipleFunction("mutation", graphqlQueryModel)

	utils.ConnectToGraphQLServer(w, graphqlQuery)
}

func UpdateProjectColor(w http.ResponseWriter, r *http.Request, projectService *services.ProjectService) {
	var body map[string]interface{}
	body, err := controller_utils.MappingBody(w, r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	projectId := chi.URLParam(r, "id")
	input := mapper.UpdateProjectColorRequestDTOMapper(body, projectId)

	graphqlQueryModel := []base_dtos.GraphQLQuery{}
	graphqlQueryModel = append(graphqlQueryModel, base_dtos.GraphQLQuery{FunctionName: "updateProjectColor", QueryInput: input, QueryOutput: model.Project{}})
	graphqlQuery := utils.GenerateGraphQLQueryWithMultipleFunction("mutation", graphqlQueryModel)

	utils.ConnectToGraphQLServer(w, graphqlQuery)
}

func ArchiveProject(w http.ResponseWriter, r *http.Request, projectService *services.ProjectService) {
	projectId := chi.URLParam(r, "id")
	input := mapper.GetId(projectId)

	graphqlQueryModel := []base_dtos.GraphQLQuery{}
	graphqlQueryModel = append(graphqlQueryModel, base_dtos.GraphQLQuery{FunctionName: "archiveProject", QueryInput: input, QueryOutput: model.Project{}})
	graphqlQuery := utils.GenerateGraphQLQueryWithMultipleFunction("mutation", graphqlQueryModel)

	utils.ConnectToGraphQLServer(w, graphqlQuery)
}

func EnableProject(w http.ResponseWriter, r *http.Request, projectService *services.ProjectService) {
	projectId := chi.URLParam(r, "id")
	input := mapper.GetId(projectId)

	graphqlQueryModel := []base_dtos.GraphQLQuery{}
	graphqlQueryModel = append(graphqlQueryModel, base_dtos.GraphQLQuery{FunctionName: "enableProject", QueryInput: input, QueryOutput: model.Project{}})
	graphqlQuery := utils.GenerateGraphQLQueryWithMultipleFunction("mutation", graphqlQueryModel)

	utils.ConnectToGraphQLServer(w, graphqlQuery)
}
